onnxruntime/core/providers/cuda/nn/rms_norm.cc

```diff
@@ -0,0 +1,105 @@
+                                    .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())    \
+                                    .TypeConstraint("V", DataTypeImpl::GetTensorType<T>()),   \
+                                RMSNorm<T, float, T, true>);
+
+REGISTER_KERNEL_TYPED(float)
+REGISTER_KERNEL_TYPED(double)
+REGISTER_KERNEL_TYPED(MLFloat16)
+REGISTER_KERNEL_TYPED(BFloat16)
+
+// The following code is shared from "core/providers/cuda/nn/layer_norm.cc".
+// It is used to implement the RMSNorm kernel, which is a simplified version of LayerNorm.
+template <typename T, typename U, typename V, bool simplified>
+RMSNorm<T, U, V, simplified>::RMSNorm(const OpKernelInfo& op_kernel_info)
+    : CudaKernel(op_kernel_info) {
+  ORT_ENFORCE(op_kernel_info.GetAttr("axis", &axis_).IsOK());
+  float tmp_epsilon;
+  ORT_ENFORCE(op_kernel_info.GetAttr<float>("epsilon", &tmp_epsilon).IsOK());
+  epsilon_ = tmp_epsilon;
+}
+
+template <typename T, typename U, typename V, bool simplified>
+Status RMSNorm<T, U, V, simplified>::ComputeInternal(OpKernelContext* ctx) const {
+  typedef typename ToCudaType<T>::MappedType CudaT;
+  typedef typename ToCudaType<U>::MappedType CudaU;
+  typedef typename ToCudaType<V>::MappedType CudaV;
+  // Inputs
+  const Tensor* X = ctx->Input<Tensor>(0);
+  const Tensor* scale = ctx->Input<Tensor>(1);
+  const Tensor* bias = ctx->Input<Tensor>(2);
+
+  auto X_data = reinterpret_cast<const CudaT*>(X->Data<T>());
+  auto scale_data = reinterpret_cast<const CudaV*>(scale->Data<V>());
+  auto bias_data = (simplified || (nullptr == bias)) ? nullptr : reinterpret_cast<const CudaV*>(bias->Data<V>());
+
+  const TensorShape& x_shape = X->Shape();
+  auto x_num_dims = x_shape.NumDimensions();
+  const int64_t axis = HandleNegativeAxis(axis_, x_num_dims);
+
+  const TensorShape& scale_shape = scale->Shape();
+  const TensorShape& bias_shape = bias_data ? bias->Shape() : TensorShape();
+
+  LayerNormParams params;
+  ORT_RETURN_IF_ERROR(LayerNormHelper::CheckInputs(x_shape, scale_shape, bias_shape, bias_data != nullptr, axis, params));
+
+  // Outputs
+  Tensor* Y = ctx->Output(0, x_shape);
+  auto Y_data = reinterpret_cast<CudaV*>(Y->MutableData<V>());
+
+  // Mean and variance
+  std::vector<int64_t> mean_inv_std_var_dim;
+  for (int i = 0; i < static_cast<int>(x_num_dims); ++i) {
+    if (i < axis) {
+      mean_inv_std_var_dim.emplace_back(x_shape.GetDims()[i]);
+    } else {
+      mean_inv_std_var_dim.emplace_back(1);
+    }
+  }
+  int output_index = 1;
+
+  CudaU* mean_data = nullptr;
+
+  if (!simplified) {
+    Tensor* mean = ctx->Output(output_index++, TensorShape(mean_inv_std_var_dim));
+    if (mean != nullptr) {
+      mean_data = reinterpret_cast<CudaU*>(mean->MutableData<U>());
+    }
+  }
+
+  CudaU* inv_var_data = nullptr;
+  Tensor* var = ctx->Output(output_index, TensorShape(mean_inv_std_var_dim));
+  if (var != nullptr) {
+    inv_var_data = reinterpret_cast<CudaU*>(var->MutableData<U>());
+  }
+
```

Comment:
There is no such output in op spec. I think these lines are not used.

Reply:
Removed